codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -560,11 +560,15 @@ public ChannelFuture goAway(final ChannelHandlerContext ctx, final int lastStrea
             future.addListener(new GenericFutureListener<ChannelFuture>() {
                 @Override
                 public void operationComplete(ChannelFuture future) throws Exception {
-                    if (!future.isSuccess()) {
+                    if (future.isSuccess()) {
+                        if (errorCode != NO_ERROR.code()) {
+                            ctx.close();
+                        }
+                    } else {
                         String msg = format("Sending GOAWAY failed: lastStreamId '%d', errorCode '%d', " +
                                             "debugData '%s'.", lastStreamId, errorCode, debugData);
                         logger.error(msg, future.cause());
-                        ctx.channel().close();
+                        ctx.close();
```

Comment:
The [AbstractChannelHandlerContext.close(..)](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java#L440) does a bit of extra work related to the `invoker` to ensure you are in the proper event loop and also notifies the [DefaultChannelHandlerInvoker](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/DefaultChannelHandlerInvoker.java#L276) -> [ChannelOutBoundHandler](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/ChannelHandlerInvokerUtil.java#L133).  [AbstractChannel](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/AbstractChannel.java#L199) just closes the pipeline...which then closes the [AbstractChannelHandlerContext](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java#L996). @normanmaurer - Anything else?